src/compiler/scala/tools/nsc/backend/jvm/BytecodeWriters.scala

```diff
@@ -118,10 +135,20 @@ trait BytecodeWriters {
     def writeClass(label: String, jclassName: String, jclassBytes: Array[Byte], outfile: AbstractFile) {
       assert(outfile != null,
              "Precisely this override requires its invoker to hand out a non-null AbstractFile.")
-      val outstream = new DataOutputStream(outfile.bufferedOutput)
+      if (outfile.file != null) {
+        try {
+          Files.write(outfile.file.toPath, jclassBytes)
+        } catch {
+          case _: java.nio.file.NoSuchFileException =>
```

Comment:
I found some small but measurable improvement from this approach, as there are typically many classfiles per directory, which makes the optimism pays off. I could split this into two commits so that we could reproduce that measurement.